modules/bm_sim/src/lookup_structures.cpp

```diff
@@ -0,0 +1,263 @@
+  }
+
+  void add_entry(const LPMMatchKey &key,
+                 internal_handle_t handle) override {
+    trie.insert_prefix(key.data, key.prefix_length,
+                       reinterpret_cast<uintptr_t>(handle));
+  }
+
+  void delete_entry(const LPMMatchKey &key) override {
+    trie.delete_prefix(key.data, key.prefix_length);
+  }
+
+  void clear() override {
+    trie.clear();
+  }
+
+ private:
+  LPMTrie trie;
+};
+
+class ExactMap : public ExactLookupStructure {
+ public:
+  explicit ExactMap(size_t size) {
+    entries_map.reserve(size);
+  }
+
+  bool lookup(const ByteContainer &key,
+              internal_handle_t *handle) const override {
+    const auto entry_it = entries_map.find(key);
+    if (entry_it == entries_map.end()) {
+      return false;  // Nothing found
+    } else {
+      *handle = entry_it->second;
+      return true;
+    }
+  }
+
+  bool entry_exists(const ExactMatchKey &key) const override {
+    (void) key;
+    return entries_map.find(key.data) != entries_map.end();
+  }
+
+  void add_entry(const ExactMatchKey &key,
+                 internal_handle_t handle) override {
+    entries_map[key.data] = handle;  // key is copied, which is not great
+  }
+
+  void delete_entry(const ExactMatchKey &key) override {
+    entries_map.erase(key.data);
+  }
+
+  void clear() override {
+    entries_map.clear();
+  }
+
+ private:
+  std::unordered_map<ByteContainer, internal_handle_t, ByteContainerKeyHash>
+    entries_map{};
+};
+
+class TernaryMap : public TernaryLookupStructure {
+ public:
+  explicit TernaryMap(size_t size, size_t nbytes_key)
+    : keys(size)
+    , nbytes_key(nbytes_key) {}
+
+  bool lookup(const ByteContainer &key_data,
+              internal_handle_t *handle) const override {
+    auto min_priority = std::numeric_limits<
+                          decltype(TernaryMatchKey::priority)>::max();
+    bool match;
+
+    const TernaryMatchKey *min_key = nullptr;
+    internal_handle_t min_handle = 0;
+
+    for (auto it = keys.begin(); it != keys.end(); ++it) {
+      if (it->priority >= min_priority) continue;
+
+      match = true;
+      for (size_t byte_index = 0; byte_index < nbytes_key; byte_index++) {
+        if (it->data[byte_index] !=
+            (key_data[byte_index] & it->mask[byte_index])) {
+          match = false;
+          break;
+        }
+      }
+
+      if (match) {
+        min_priority = it->priority;
+        min_key      = &*it;
+        min_handle   = std::distance(keys.begin(), it);
+      }
+    }
+
+    if (min_key) {
+      *handle = min_handle;
+      return true;
+    }
+
+    return false;
+  }
+
+  bool entry_exists(const TernaryMatchKey &key) const override {
+    auto handle = find_handle(key);
+
+    return handle != keys.size();
+  }
+
+  void add_entry(const TernaryMatchKey &key,
+                 internal_handle_t handle) override {
+    assert(handle < keys.size());
+
+    keys[handle] = key;
+  }
+
+  void delete_entry(const TernaryMatchKey &key) override {
```

Comment:
I agree - one thought I had as I was writing this was to keep the keys sorted by priority so that we can at least return on the first match instead of having to scan the entire table even after having matched. This would then require storing the handles along with the keys as they would no longer work as indices.

In any case, and as you said, I figured I'd leave it as it was for now.